### DISCUSSION TITLE
Submitting DIRAC jobs via HTTP

### DISCUSSION THREAD

**Discussion by kreczko** [DIRACGrid/DIRAC#6121] [Q&A] [answered]
At the last DIRAC users' workshop two teasers were shown for HTTP services:
- [Overview when more services will be introduced](https://indico.cern.ch/event/1107386/contributions/4827827/attachments/2439420/4178399/Intro%20%284%29.pdf)
- [More info and example with FileCatalog](https://indico.cern.ch/event/852597/contributions/4331720/attachments/2241040/3799728/HttpsInDIRAC.pdf)

But even given that information (and release notes for 7.3 and 8.0) as well as the documentation, I am still lacking.
How can I submit and manage jobs via HTTP? (7.3 mentions `JobManager` - is that it?)
If it is not possible yet, is it on the roadmap?

BTW: Do you have a secret URL for each DIRAC server that has descriptions of the HTTP API it supports? (e.g. like automatically done by [FastAPI](https://www.linode.com/docs/guides/documenting-a-fastapi-app-with-openapi/example_doc_image_3_hufd23c799025bb51d1b2172c48ed14774_149996_1600x0_resize_q71_bgfafafc_catmullrom_2.jpg))

Context: I would like to make a "zero dependency" DIRAC client

**Comment by fstagni** [comment]
Hi, 
on the client side it would be something like what is described in slide 6 of https://indico.cern.ch/event/852597/contributions/4331720/attachments/2241040/3799728/HttpsInDIRAC.pdf but with `url=https://server:8443/WorkloadManagement/TornadoJobManager` and `kwargs={'method': 'submitJob'}` (for ref the [code](https://github.com/DIRACGrid/DIRAC/blob/integration/src/DIRAC/WorkloadManagementSystem/Service/JobManagerHandler.py#L126)).
In this context there's also a `TornadoJobMonitoring` service for https based monitoring.

This of course assumes that the server you are submitting to has the TornadoJobManager service up and running, and the rights to submit.

**Comment by marianne013** [comment]
Luke is talking about the GridPP one, which is still on v7r2 and doesn't have the TornadoJobManager.
Luke is a member of dteam (I assume) - couldn't he just test against the certification server for a proof of principle ? It's basically DIRAC code development he's suggesting :-D
I talked to Simon briefly about Tornado and he mumbled something about "eventually" - our next big step would be to upgrade our production server to v7r3 which I think does Tornado in principle, though we'd always start with our pre-prod servers if we wanted to look at this in earnest.

**Reply by fstagni** [reply]
Yes, if Luke is a member of dteam, then he can use the DIRAC certification setup for test. If that's the case you can guide him through the various links. You would need to add him manually to the CS/Registry of lbcertifdirac70.

**Comment by marianne013** [comment]
@kreczko Let me know if you wanted to proceed with this.

**Reply by kreczko** [reply]
Yes. Sorry for the late reply.

**Reply by marianne013** [reply]
@kreczko You should now be able to use the certification server (https://lbcertifdirac70.cern.ch/DIRAC/) under the dteam and gridpp VOs.
Could you have a look please and let me know if there is an issue ?

**Reply by kreczko** [reply]
Something certainly changed

## Before
![image](https://user-images.githubusercontent.com/1213276/173607973-689d5132-73db-42ba-9d2c-efccab0e8fb0.png)

## After
![image](https://user-images.githubusercontent.com/1213276/173608059-f7e6247f-d7bb-4d26-be19-48ef9eebf66b.png)

I will try an HTTP request next

**Reply by kreczko** [reply]
When trying to post to https://lbcertifdirac70.cern.ch:8443/WorkloadManagement/TornadoJobMonitoring
I get a 404 error - not found.

@fstagni What kind of queries can I do against this server?

**Reply by fstagni** [reply]
The URL is https://lbcertifdirac70.cern.ch:8443/WorkloadManagement/JobMonitoring
(no "TornadoJobMonitoring" but "JobMonitoring)

The RPC are the "export_" methods in https://github.com/DIRACGrid/DIRAC/blob/integration/src/DIRAC/WorkloadManagementSystem/Service/JobMonitoringHandler.py

You can try by calling `getJobs`

**Reply by kreczko** [reply]
Thanks, that worked:
```python
import requests

server = "lbcertifdirac70.cern.ch"
url=f"https://{server}:8443/WorkloadManagement/JobMonitoring"
capath="/etc/grid-security/certificates"
cert="/tmp/x509up_u1000"
params = {
    "method": "getJobs"
}

with requests.post(url, data=params, cert=cert, verify=capath) as r:
    print(r.text)
```

**Reply by kreczko** [reply]
hmm, if I try to submit a job, I get error 400: Bad Request.
Code:
```python
import requests

jdl="""
JobName = "DIRAC_lkreczko_642778";
Executable = "/bin/ls";
Arguments = "-ltrA";
StdOutput = "std.out";
StdError = "std.err";
OutputSandbox = {"std.out","std.err"};
"""

def _query(url, params):
    capath="/etc/grid-security/certificates"
    cert="/tmp/x509up_u1000" # TODO: default path should be /tmp/x509up_u<user id>
    with requests.post(url, data=params, cert=cert, verify=capath) as r:
        return r.text


def submit_job(server, jdl):
    url=f"https://{server}:8443/WorkloadManagement/JobManager"
    params = {
        "method": "submitJob",
        "jobDesc": jdl
    }
    return _query(url, params)


if __name__ == "__main__":
    server = "lbcertifdirac70.cern.ch"
    result = submit_job(server, jdl)
    # result = get_jobs(server)
    # result = get_files(server)
    print(result)
```

Which (to me) looks like it should work: https://github.com/DIRACGrid/DIRAC/blob/integration/src/DIRAC/WorkloadManagementSystem/Service/JobManagerHandler.py#L126

**Reply by kreczko** [reply]
The `whoami` example works well:
```python
def whoami(server):
    url=f"https://{server}:8443/DataManagement/FileCatalog"
    params = {
        "method": "whoami"
    }
    return _query(url, params)
```

```
{"OK": true, "Value": {"subject": "/C=UK/O=eScience/OU=Bristol/L=IS/CN=lukasz kreczko/CN=3866680583", "issuer": "/C=UK/O=eScience/OU=Bristol/L=IS/CN=lukasz kreczko", "secondsLeft": 40513, "isProxy": true, "isLimitedProxy": false, "validDN": false, "validGroup": false, "DN": "/C=UK/O=eScience/OU=Bristol/L=IS/CN=lukasz kreczko", "identity": "/C=UK/O=eScience/OU=Bristol/L=IS/CN=lukasz kreczko", "group": "dteam_user", "properties": ["NormalUser"], "username": "lkreczko"}}
```

**Reply by fstagni** [reply]
```
2022-06-14 15:36:31 UTC Tornado/Tornado/TornadoJobManagerHandler [140319882528512] NOTICE: Incoming request (137.222.58.191:27424)[dteam_user:lkreczko] WorkloadManagement/JobManager: submitJob
2022-06-14 15:36:31 UTC ExternalLibrary/stomp.py DEBUG: Sending frame: [b'SEND', b'\n', b'content-length:496\n', b'destination:/queue/dirac.logging\n', b'\n', b'{"levelname": "NOTICE", "message": "Incoming request", "asctime": "2022-06-14 15:36:31,172098", "componentname": "Tornado/Tornado", "varmessage": "(137.222.58.191:27424)[dteam_user:lkreczko] WorkloadManagement/JobManager: submitJob", "spacer": " ", "customname": "/TornadoJobManagerHandler", "headerIsShown": true, "timeStampIsShown": true, "contextIsShown": true, "threadIDIsShown": true, "color": true, "tornadoComponent": "WorkloadManagement/JobManager", "hostname": "lbcertifdirac70.cern.ch"}', b'\x00']
2022-06-14 15:36:31 UTC Tornado/Tornado/TornadoJobManagerHandler [140319882528512] ERROR: Exception serving request HTTP 400: Bad Request (Missing argument clientSetup):MissingArgumentError()
Traceback (most recent call last):
  File "/opt/dirac/versions/v8.0.0a22-1652335739/Linux-x86_64/lib/python3.9/site-packages/DIRAC/Core/Tornado/Server/private/BaseRequestHandler.py", line 646, in __executeMethod
    self.initializeRequest()
  File "/opt/dirac/versions/v8.0.0a22-1652335739/Linux-x86_64/lib/python3.9/site-packages/DIRAC/WorkloadManagementSystem/Service/TornadoJobManagerHandler.py", line 16, in initializeRequest
    self.diracSetup = self.get_argument("clientSetup")
  File "/opt/dirac/versions/v8.0.0a22-1652335739/Linux-x86_64/lib/python3.9/site-packages/tornado/web.py", line 407, in get_argument
    return self._get_argument(name, default, self.request.arguments, strip)
  File "/opt/dirac/versions/v8.0.0a22-1652335739/Linux-x86_64/lib/python3.9/site-packages/tornado/web.py", line 484, in _get_argument
    raise MissingArgumentError(name)
tornado.web.MissingArgumentError: HTTP 400: Bad Request (Missing argument clientSetup)
```

I am not sure here. A `"clientSetup": "DIRAC-Certification"` must be present somehow. How did you create the proxy?
@chaen might know better.

**Reply by kreczko** [reply]
Created proxy via `voms-proxy-init --voms dteam`.

Tried something more simple (no extra parameters):
```python
def getMaxParametricJobs(server):
    url=f"https://{server}:8443/WorkloadManagement/JobManager"
    params = {
        "method": "getMaxParametricJobs"
    }
    return _query(url, params)
```
&rarr; also gets me a bad request. So something is very different with the `WorkloadManagement/JobManager` endpoint compared to the `WorkloadManagement/JobMonitoring` and `DataManagement/FileCatalog` endpoints.

BTW: The [REST API](https://dirac.readthedocs.io/en/latest/DeveloperGuide/REST/index.html) looks much simpler. Is that available?

**Reply by chaen** [reply]
you shouldn't use `voms-proxy-init` as it doesn't know anything about DIRAC so can't set the DIRAC options in the proxy. Use `dirac-proxy-init`.

**Reply by fstagni** [reply]
Or you can use `dirac-login`, which for this version is equivalent to `dirac-proxy-init`. 

The REST API is not really tested. You can try but I would not expect great results.

**Reply by marianne013** [reply]
The whole idea was not to have to install a DIRAC UI, so using dirac* commands somewhat violates the spirit of the idea.

**Reply by chaen** [reply]
In principle, it should work with just the certificate. Or rather, in v7r3, it should work with just the certificate (can't comment on v8). But I remember the `jobManager` was special, but only @fstagni can comment on why

**Reply by fstagni** [reply]
JobManager isn't particularly special. 

But I am afraid that to get to a "no-DIRAC" client, you would need to wait at least https://github.com/DIRACGrid/DIRAC/pull/6164

**Reply by kreczko** [reply]
Update on this: I've previously used dirac-login on a different machine, then copied the proxy onto mine. That seemed to work.
We finally have a person tackling this issue on a regular basis (instead of me in my spare time): @seriksen.

Sam just applied for `dteam` and `gridpp` membership and should be able to talk the the certification server once that is done.

Have there been any significant, related changes since? I see the mentioned PR has been merged.

**Reply by fstagni** [reply]
Hello, the mentioned PR has been merged, yes, as well some other fixes going into the development versions of DIRAC (8.1 -- production is 8.0). There is also an issue https://github.com/DIRACGrid/DIRAC/issues/6496 for following up. As you could read in it there is at least an issue, that could be addressed by @seriksen if he wants to dive deep in the DIRAC code.

For a bit more elaborated roadmap on this maybe you can connect to our bi-weekly meeting.

**Reply by seriksen** [reply]
@marianne013 can I be added to the CS/Registry of lbcertifdirac70, please

**Reply by marianne013** [reply]
For this I (or@fstagni) would need your DN and CA, please. I can't find anyone called 'eriksen' in the gridpp VO, so I assume you just want to stick with dteam ?

**Reply by seriksen** [reply]
I'm still waiting for gridpp to be approved.

`CN=sam eriksen, L=IS, OU=Bristol, O=eScience, C=UK`

**Reply by marianne013** [reply]
@seriksen This should now all be done (incl gridpp)

**Reply by seriksen** [reply]
Possibly a question for @fstagni , I've tried submitting the job above: https://github.com/DIRACGrid/DIRAC/discussions/6121#discussioncomment-2948494
but when I view it on https://lbcertifdirac70.cern.ch/DIRAC/ it fails with `Failed to setup proxy: Error retrieving proxy`. Is there something that I'm missing from the job parameters?

**Reply by seriksen** [reply]
I don't know if this is related, but when running `dirac-proxy-init`, I am getting that I am not registered;

```
sam@dell:~/DIRAC/https_dirac/diracos$ dirac-proxy-init -d
Generating proxy...
Enter Certificate password: 
Contacting CS...
Could not get a default group for DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen: No username found for dn /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
Checking DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
No username found for dn /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen is not registered
```

**Reply by fstagni** [reply]
Try with `dirac-proxy-init -g dteam_user` (or `dirac-login dteam_user`) and let us know.

**Reply by seriksen** [reply]
```
(base) sam@dell:~/DIRAC/https_dirac/diracos$ dirac-proxy-init -g dteam_user 
Generating proxy...
Enter Certificate password: 
DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen is not registered
```

```
(base) sam@dell:~/DIRAC/https_dirac/diracos$ dirac-login dteam_user
Enter Certificate password: 
Unauthorized query ( 1111 : Unauthorized query)
```

```
(base) sam@dell:~/DIRAC/https_dirac/diracos$ dirac-proxy-info
subject      : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen/CN=2126418691/CN=770332529/CN=7566964172
issuer       : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen/CN=2126418691/CN=770332529
identity     : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
timeleft     : 11:49:02
path         : /tmp/x509up_u1000
VOMS         : True
VOMS fqan    : ['/dteam']
```

**Reply by seriksen** [reply]
Do you have any suggestions?

**Reply by andresailer** [reply]
Run `dirac-proxy-init -g dteam_user -ddd` (debug flags!) and see if that actually talks to the correct configuration service.

**Reply by seriksen** [reply]
On dteam:
```
(base) sam@dell:~/DIRAC/dask-dirac$ dirac-proxy-init -g dteam_user -ddd
2023-01-31 15:44:50 UTC Framework [140502386767680] DEBUG: dirac.cfg should be at /home/sam/DIRAC/diracod2/diracos/etc/dirac.cfg
2023-01-31 15:44:50 UTC Framework [140502386767680] DEBUG: CFG merged
2023-01-31 15:44:50 UTC Framework [140502386767680] DEBUG: Updating configuration internals
2023-01-31 15:44:50 UTC Framework [140502386767680] DEBUG: Updating configuration internals
Parsing command line
Trying to load Resources.LogBackends.StdoutBackend
Trying to load DIRAC.Resources.LogBackends.StdoutBackend
2023-01-31 15:44:50 UTC dirac-proxy-init [140502386767680] DEBUG: Updating configuration internals
2023-01-31 15:44:50 UTC dirac-proxy-init [140502386767680] NOTICE: Generating proxy...
Enter Certificate password: 
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] INFO: Contacting CS...
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG: Refreshing configuration...
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG: Refreshing from list ['dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server']
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG: Randomized server list is dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG:  Trying to refresh from dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:51 UTC dirac-proxy-init/DIRAC.Core.Tornado.Client.ClientSelector [140502386767680] DEBUG: Trying to autodetect client for dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG: Already given a valid url dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:51 UTC dirac-proxy-init [140502386767680] DEBUG: Trying to connect to: dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:52 UTC dirac-proxy-init [140502386767680] DEBUG: Connected to: dips://dirac01.grid.hep.ph.ic.ac.uk:9135/Configuration/Server
2023-01-31 15:44:52 UTC dirac-proxy-init [140502386767680] DEBUG: New connection -> 146.179.232.10:9135
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] DEBUG: New version available Updating to version 2023-01-31 15:41:35.644166...
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] DEBUG: Updating configuration internals
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] DEBUG: Updated to version 2023-01-31 15:41:35.644166
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] DEBUG: Trying to load Resources.LogBackends.StdoutBackend
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] DEBUG: Trying to load DIRAC.Resources.LogBackends.StdoutBackend
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] INFO: Checking DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] INFO: Username is sam.eriksen
2023-01-31 15:44:54 UTC dirac-proxy-init [140502386767680] ERROR: Requested group dteam_user is not valid for DN /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
```
whilst on gridpp
```
(base) sam@dell:~/DIRAC/dask-dirac$ dirac-proxy-init -g gridpp_user
Generating proxy...
Enter Certificate password: 
Added VOMS attribute /gridpp
Uploading proxy..
Function requestDelegationUpload expects at least 1 arguments
Proxy generated:
subject      : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen/CN=1999752188/CN=508312065
issuer       : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen/CN=1999752188
identity     : /C=UK/O=eScience/OU=Bristol/L=IS/CN=sam eriksen
timeleft     : 23:53:58
DIRAC group  : gridpp_user
path         : /tmp/x509up_u1000
username     : sam.eriksen
properties   : NormalUser, JobSharing, AccountingMonitor
VOMS         : True
VOMS fqan    : ['/gridpp']
```

**Reply by andresailer** [reply]
Please give the output from your installation in  `~/DIRAC/https_dirac/diracos`

**Comment by marianne013** [comment]
@seriksen 
There has been a bit of a mix up.
Your UI (? - I am not sure how you set this up when you use http) seems to be poiting at the GridPP production (dirac01.grid....) server which does not have a version that supports jobsubmission via http (and does not support dteam). 
I think the intention was to test against the DIRAC certification server (as this will have the latest release on), which can be found at: 
https://lbcertifdirac70.cern.ch/DIRAC/

**Reply by seriksen** [reply]
I ran `dirac-configure` first which set 
```
Configuration is:
  * Setup: DIRAC-Certification
  * Configuration server: https://lbcertifdirac70.cern.ch:9135/Configuration/Server
```
which is correct, no?

**Reply by kreczko** [reply]
We've been trying to find ways to connect to the GridPP diracdev and certifdirac.

For certif, we tried
```
dirac-configure
```
and selected the certifdirac - is that not enough?

Can you please point us to documentation on how to choose the target dirac instance?

For our submission tests, we give different URLs (essentially a curl):

```
def submit_job(settings: DiracSettings, jdl: str) -> Any:
    """Submit a job to a DIRAC server"""
    endpoint = "WorkloadManagement/JobManager"
    settings.query_url = f"{settings.server_url}/{endpoint}"
    params = {"method": "submitJob", "jobDesc": jdl}
    return _query(settings, params)
```
&rarr; always querying lbcertifdirac.

@fstagni we are also getting
```
requests.exceptions.HTTPError: 500 Server Error: Internal Server Error for url: https://lbcertifdirac70.cern.ch:8443/WorkloadManagement/JobManager

```

with the above.
What format does the query needs to be at?
```python
params = {"method": "submitJob", "jobDesc": jdl}
```
Is the method OK? What format should `jdl` be? A str, dictionary or `json.dumps`?

**Comment by chaen** [comment, marked as answer]
`dirac-configure` won't overwrite an existing `dirac.cfg`, so make sure you don't have one in place to start with.

**Reply by seriksen** [reply]
thanks, that's solved it. working now

**Reply by chaen** [reply]
That should help in the future
https://github.com/DIRACGrid/DIRAC/pull/6764

**Comment by seriksen** [comment]
Coming back to what @kreczko posted [here](https://github.com/DIRACGrid/DIRAC/discussions/6121#discussioncomment-4831591), is there an example of a submission test anywhere?

**Reply by fstagni** [reply]
I usually test with 

```python
import json
import requests

jdl="""
JobName = "DIRAC_no-client-test";
Executable = "/bin/ls";
Arguments = "-ltrA";
StdOutput = "std.out";
StdError = "std.err";
OutputSandbox = {"std.out","std.err"};
Owner = "fstagni";
OwnerGroup = "dteam_user";
"""

def _query(url, params):
    capath="/etc/grid-security/certificates"
    cert="/tmp/x509up_u49429"
    with requests.post(url, data=params, cert=cert, verify=capath) as r:
        return r.text


def submit_job(server, jdl):
    url=f"https://{server}:8443/WorkloadManagement/JobManager"
    params = {
        "method": "submitJob",
        "args": json.dumps([jdl])
    }
    return _query(url, params)

def get_jobs(server):
  url=f"https://{server}:8443/WorkloadManagement/JobMonitoring"
  params = {
      "method": "getJobs"    
  }
  return _query(url, params)


if __name__ == "__main__":
    server = "lbcertifdirac70.cern.ch"
    #result = submit_job(server, jdl)
    result = get_jobs(server)
    # result = get_files(server)
    print(result)
```

You should of course set your "Onwer" and "OwnerGroup"

**Reply by kreczko** [reply]
Aha! So that's the secret:
```
"args": json.dumps([jdl])
```

We've been also trying with
```python
jdl = {
    "JobName": "DASK_DIRAC",
    "Executable": "/bin/ls",
    "Arguments": "-ltrA",
    "StdOutput": "std.out",
    "StdError": "std.err",
    "OutputSandbox": ["std.out","std.err"]
}
```
:D

OK, your example looks very doable, thanks!

**Comment by seriksen** [comment]
Hello, We're currently waiting to be able to submit jobs via HTTP to https://diracdev.grid.hep.ph.ic.ac.uk:8443/DIRAC/. 
Whilst that is happening, is it possible to add Bristol (UKI-SOUTHGRID-BRIS-HEP) as a site to lbcertifdirac70 whilst we wait for that to happen? Not sure how big a task that is?

**Reply by fstagni** [reply]
Hi, I just bring back this old discussion, that I've got reminded somehow today, to say that all http developments for DIRAC stopped, and are all diverted to [diracx](https://github.com/DIRACGrid/diracx)
